src/lib/logs/logs_processor.py

```diff
@@ -14,51 +14,141 @@
 
 import json
 import queue
-import time
 from datetime import datetime, timezone
-from queue import Queue
-from typing import Optional, Dict
+from asyncio import Queue
+from typing import Optional, Dict, NamedTuple, List
 
+import base64
 from dateutil.parser import *
 from google.pubsub_v1 import ReceivedMessage, PubsubMessage
 
 from lib.context import LogsProcessingContext
-from lib.logs.log_forwarder_variables import EVENT_AGE_LIMIT_SECONDS, CONTENT_LENGTH_LIMIT, \
-    ATTRIBUTE_VALUE_LENGTH_LIMIT, DYNATRACE_LOG_INGEST_CONTENT_MARK_TRIMMED, CLOUD_LOG_FORWARDER, \
-    CLOUD_LOG_FORWARDER_POD
-from lib.logs.log_self_monitoring import LogSelfMonitoring, put_sfm_into_queue
+from lib.logs.log_forwarder_variables import (
+    EVENT_AGE_LIMIT_SECONDS,
+    CONTENT_LENGTH_LIMIT,
+    ATTRIBUTE_VALUE_LENGTH_LIMIT,
+    DYNATRACE_LOG_INGEST_CONTENT_MARK_TRIMMED,
+    CLOUD_LOG_FORWARDER,
+    CLOUD_LOG_FORWARDER_POD,
+    REQUEST_BODY_MAX_SIZE,
+    REQUEST_MAX_EVENTS,
+)
+from lib.logs.log_self_monitoring import LogSelfMonitoring, aggregate_self_monitoring_metrics, put_sfm_into_queue
 from lib.logs.metadata_engine import MetadataEngine, ATTRIBUTE_CONTENT, ATTRIBUTE_TIMESTAMP
 
 _metadata_engine = MetadataEngine()
 
 
 class LogProcessingJob:
-    payload: str
     self_monitoring: LogSelfMonitoring
 
-    def __init__(self, payload: str, self_monitoring: LogSelfMonitoring):
+    def __init__(self, payload: str, self_monitoring: LogSelfMonitoring, ack_id):
         self.payload = payload
         self.self_monitoring: LogSelfMonitoring = self_monitoring
         self.bytes_size = len(payload.encode("UTF-8"))
+        self.ack_id = ack_id
 
 
-def _prepare_context_and_process_message(sfm_queue: Queue, message: ReceivedMessage) -> Optional[LogProcessingJob]:
+class LogBatch(NamedTuple):
+    serialized_batch: str
+    number_of_logs_in_batch: int
+    ack_ids: List[str]
+    size_batch_bytes: int
+    self_monitoring: LogSelfMonitoring
+
+
+
+def prepare_batches(logs: List[LogProcessingJob]) -> List[LogBatch]:
+    batches: List[LogBatch] = []
+
+    logs_for_next_batch: List[str] = []
+    ack_ids_for_next_batch: List[str] = []
+    logs_for_next_batch_total_len = 0
+    logs_for_next_batch_events_count = 0
+
+    log_entries = 0
+    batch_sfm_monitoring_list = []
+    for log_entry in logs:
+        new_batch_len = (
+            logs_for_next_batch_total_len + 2 + len(logs_for_next_batch) - 1
```

Comment:
The comment is great, but I thing that assigning a variable like "bracket_length = 2" would be even better 😄